### PR TITLE
Add argument `none_value` for `None` representation in loading and dumping

### PR DESCRIPTION
**Pull request by pwwang**
Related: https://github.com/samuelcolvin/rtoml/issues/23, https://github.com/toml-lang/toml/issues/30

- load/loads
  ```python
  # By default, nothing should be loaded as None, since it doesn't exist in TOML
  rtoml.loads('x = "null"') == {'x': "null"}
  rtoml.loads('x = ""') == {'x': ""}
  # Now with the PR, we can tell what to be interpreted as None
  rtoml.loads('x = "null"', none_value="null") == {'x': None}
  rtoml.loads('x = ""', none_value="") == {'x': None}
  ```
- dump/dumps
  ```python
  # By default, None dumps into "null"
  rtoml.dumps({"x": None}) == 'x = "null"\n'
  # Now with the PR, we can control what None should be dumped into
  rtoml.dumps({"x": None}, none_value="@None") == 'x = "@None"\n'
  ```

Reproducibility between loading and dumping using the same `none_value`:

```python
s = 'x = "py:None"\n'
rtoml.loads(s, none_value="py:None") == {'x': None}
rtoml.dumps({'x': None}, none_value="py:None") == s

s = {'x': None}
rtoml.dumps(s, none_value="py:None") == 'x = "py:None"\n'
rtoml.loads('x = "py:None"\n', none_value="py:None") == s
```

The PR modifies the visitor at rust end, and no re-visit is needed to covert the values at python end (like suggested by #23). So it barely affects the speed.

I think the only "drawback" is that only string `None`-representation is supported. You can't convert `0` (integer) to `None`, for example, while loading, or vice versa while dumping.

You can also omit `None` values or items that are associated with `None` value with `omit_none=True`. See https://github.com/samuelcolvin/rtoml/issues/23#issuecomment-917640979

```python
rtoml.dumps({'test': None}, omit_none=True) == ''
rtoml.dumps({'test': [1, None, 2]}, omit_none=True) == 'test = [1, 2]\n'
rtoml.dumps({'test': (1, None, 2)}, omit_none=True) == 'test = [1, 2]\n'
rtoml.dumps({'test': {'x': [None, {'y': [1, None, 2]}], 'z': None}}, omit_none=True)  # gives
"""
[[test.x]]
y = [1, 2]
"""
# But with omit_none=False
rtoml.dumps({'test': {'x': [None, {'y': [1, None, 2]}], 'z': None}}, omit_none=False, pretty=True)   # gives
"""
[test]
z = 'null'

[[test.x]]
y = [
    1,
    'null',
    2,
]
```

### CODE REVIEW

**Comment by samuelcolvin** [comment]
Looking good, suggestions:
* Rename the argument to `none_value`
* make `none_value` optional, default `None` when deserialising

**Comment by pwwang** [comment]
Revised.

**Comment by samuelcolvin** [comment]
This looks awesome, I'll review properly when I'm at a computer.

File: src/lib.rs [hunk 1]
```diff
@@ -20,25 +20,25 @@ create_exception!(_rtoml, TomlParsingError, PyValueError);
...
+    let seed = de::PyDeserializer::new(py, none_value);
     seed.deserialize(&mut deserializer)
         .map_err(|e| TomlParsingError::new_err(e.to_string()))
 }
 
 #[pyfunction]
-fn serialize(py: Python, obj: &PyAny) -> PyResult<String> {
-    let s = SerializePyObject::new(py, obj);
+fn serialize(py: Python, obj: &PyAny, none_value: String) -> PyResult<String> {
+    let s = SerializePyObject::new(py, obj, &none_value);
```
Comment (samuelcolvin) [inline src/lib.rs:32]:
```suggestion
fn serialize(py: Python, obj: &PyAny, none_value: &str) -> PyResult<String> {
    let s = SerializePyObject::new(py, obj, none_value);
```

Hopefully avoid one allocation.

File: src/lib.rs [hunk 2]
```diff
@@ -20,25 +20,25 @@ create_exception!(_rtoml, TomlParsingError, PyValueError);
...
         Ok(s) => Ok(s),
         Err(e) => Err(TomlSerializationError::new_err(e.to_string())),
     }
 }
 
 #[pyfunction]
-fn serialize_pretty(py: Python, obj: &PyAny) -> PyResult<String> {
-    let s = SerializePyObject::new(py, obj);
+fn serialize_pretty(py: Python, obj: &PyAny, none_value: String) -> PyResult<String> {
+    let s = SerializePyObject::new(py, obj, &none_value);
```
Comment (samuelcolvin) [inline src/lib.rs:41]:
```suggestion
fn serialize_pretty(py: Python, obj: &PyAny, none_value: &str) -> PyResult<String> {
    let s = SerializePyObject::new(py, obj, none_value);
```

File: src/ser.rs [hunk 3]
```diff
@@ -11,22 +11,25 @@ use crate::py_type::PyTypeLookup;
...
             ob_type_lookup: PyTypeLookup::cached(py),
         }
     }
 
-    fn with_obj(&self, obj: &'py PyAny) -> Self {
+    fn with_obj(&self, obj: &'py PyAny, none_value: &'py str) -> Self {
         Self {
             obj,
             py: self.py,
+            none_value,
```
Comment (samuelcolvin) [inline src/ser.rs:32]:
```suggestion
    fn with_obj(&self, obj: &'py PyAny) -> Self {
        Self {
            obj,
            py: self.py,
            none_value: self.none_value,
```

Then remove the argument in below calls.

File: src/ser.rs [hunk 4]
```diff
@@ -89,32 +92,32 @@ impl<'py> Serialize for SerializePyObject<'py> {
             let mut map = serializer.serialize_map(Some(len))?;
             for (k, v) in simple_items {
                 let key = table_key(k)?;
-                let value = self.with_obj(v);
+                let value = self.with_obj(v, self.none_value);
```
Comment (samuelcolvin) [inline src/ser.rs:95]:
```suggestion
                let value = self.with_obj(v);
```

File: src/ser.rs [hunk 5]
```diff
@@ -89,32 +92,32 @@ impl<'py> Serialize for SerializePyObject<'py> {
...
                 map.serialize_entry(key, &value)?;
             }
             for (k, v) in array_items {
                 let key = table_key(k)?;
-                let value = self.with_obj(v);
+                let value = self.with_obj(v, self.none_value);
```
Comment (samuelcolvin) [inline src/ser.rs:100]:
```suggestion
                let value = self.with_obj(v);
```

File: src/ser.rs [hunk 6]
```diff
@@ -89,32 +92,32 @@ impl<'py> Serialize for SerializePyObject<'py> {
...
                 map.serialize_entry(key, &value)?;
             }
             for (k, v) in dict_items {
                 let key = table_key(k)?;
-                let value = self.with_obj(v);
+                let value = self.with_obj(v, self.none_value);
```
Comment (samuelcolvin) [inline src/ser.rs:105]:
```suggestion
                let value = self.with_obj(v);
```

File: src/ser.rs [hunk 7]
```diff
@@ -89,32 +92,32 @@ impl<'py> Serialize for SerializePyObject<'py> {
...
                 map.serialize_entry(key, &value)?;
             }
             map.end()
         } else if ob_type == lookup.list {
             let py_list: &PyList = self.obj.cast_as().map_err(map_py_err)?;
             let mut seq = serializer.serialize_seq(Some(py_list.len()))?;
             for element in py_list {
-                seq.serialize_element(&self.with_obj(element))?
+                seq.serialize_element(&self.with_obj(element, self.none_value))?
```
Comment (samuelcolvin) [inline src/ser.rs:113]:
```suggestion
                seq.serialize_element(&self.with_obj(element))?
```

File: src/ser.rs [hunk 8]
```diff
@@ -89,32 +92,32 @@ impl<'py> Serialize for SerializePyObject<'py> {
...
             }
             seq.end()
         } else if ob_type == lookup.tuple {
             let py_tuple: &PyTuple = self.obj.cast_as().map_err(map_py_err)?;
             let mut seq = serializer.serialize_seq(Some(py_tuple.len()))?;
             for element in py_tuple {
-                seq.serialize_element(&self.with_obj(element))?
+                seq.serialize_element(&self.with_obj(element, self.none_value))?
```
Comment (samuelcolvin) [inline src/ser.rs:120]:
```suggestion
                seq.serialize_element(&self.with_obj(element))?
```

File: tests/test_benchmarks.py [hunk 9]
```diff
@@ -127,11 +127,16 @@ def test_data():
...
 
 @pytest.mark.benchmark(group='load-data.toml')
-def test_load_data_toml(benchmark):
+def test_loads_data_toml(benchmark):
     data_toml_str = data_toml_path.read_text()
     # benchmark(rtoml.loads, data_toml_bytes, name='loads-bytes')
-    benchmark(rtoml.loads, data_toml_str, name='loads-str')
-    benchmark(rtoml.load, data_toml_path, name='load-path')
+    benchmark(rtoml.loads, data_toml_str)
+    # benchmark(rtoml.load, data_toml_path, name='load-path')
```
Comment (samuelcolvin) [inline tests/test_benchmarks.py:134]: if you want to change these, please remove the commented out lines.

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM.

File: tests/test_benchmarks.py (hunk 9 above)
Comment (pwwang) [reply tests/test_benchmarks.py:134]: I didn't mean to change it, but it kept complaining that `rtoml.load(s)`  doesn't have `name` keyword argument. Which pytest-benchmark were you using? I don't find it in requirements.txt or test-requirements.txt.

Comment (samuelcolvin) [reply tests/test_benchmarks.py]: it's `pytest-speed` which is in requirements, pytest-speed is mostly backwards compatible with pytest-benchmark, except it allow multiple benchmarks per test - like this.

**Comment by pwwang** [comment]
The `omit_none` argument was added at https://github.com/samuelcolvin/rtoml/pull/53/commits/3be85862ca75ae880ad307aceaaa5ed1c4763fce. 

I know it's a sort of different functionality but it costs tiny modifications on the current code base. I can submit a separate PR if you want.

File: rtoml/__init__.py [hunk 10]
```diff
@@ -23,19 +23,19 @@ def load(toml: Union[str, Path, TextIO]) -> Dict[str, Any]:
...
     Parse a TOML string and return a python dict. (provided to match the interface of `json` and similar libraries)
     """
     if not isinstance(toml, str):
         raise TypeError(f'invalid toml input, must be str not {type(toml)}')
-    return _rtoml.deserialize(toml)
+    return _rtoml.deserialize(toml, none_value=none_value)
 
 
-def dumps(obj: Any, *, pretty: bool = False) -> str:
+def dumps(obj: Any, *, pretty: bool = False, none_value: str = "null", omit_none: bool = False) -> str:
```
Comment (samuelcolvin) [inline rtoml/__init__.py:38]:
```suggestion
def dumps(obj: Any, *, pretty: bool = False, none_value: Optional[str, None] = "null") -> str:
```

I'm fine with adding omit functionality here, but at least in the python interface, it should be defined via the same parameter.

If you think `none_value=None` to mean "omit `None` values" is not clean enough, I'd be fine with a sentinel value, like:

```py
class OmitType:
    pass

Omit = OmitType()

... none_value: Union[str, OmitType] = 'null'
```

File: src/ser.rs [hunk 11]
```diff
@@ -78,7 +84,9 @@ impl<'py> Serialize for SerializePyObject<'py> {
 
             for (k, v) in py_dict {
                 let v_ob_type = v.get_type_ptr() as usize;
-                if v_ob_type == lookup.dict {
+                if v_ob_type == lookup.none && self.omit_none {
```
Comment (samuelcolvin) [inline src/ser.rs:87]:
```suggestion
                if self.omit_none && (v_ob_type == lookup.none || k.is_none()) {
```
I guess `self.omit_none` is more likely to be false.

File: src/ser.rs [hunk 12]
```diff
@@ -90,30 +98,50 @@ impl<'py> Serialize for SerializePyObject<'py> {
             for (k, v) in simple_items {
                 let key = table_key(k)?;
                 let value = self.with_obj(v);
+                let v_ob_type = v.get_type_ptr() as usize;
+                if v_ob_type == lookup.none && self.omit_none {
+                    continue;
+                }
```
Comment (samuelcolvin) [inline src/ser.rs:104]: this isn't required, `None` values are already excluded.

File: src/ser.rs [hunk 13]
```diff
@@ -90,30 +98,50 @@ impl<'py> Serialize for SerializePyObject<'py> {
...
                 map.serialize_entry(key, &value)?;
             }
             for (k, v) in dict_items {
                 let key = table_key(k)?;
                 let value = self.with_obj(v);
+                let v_ob_type = v.get_type_ptr() as usize;
+                if v_ob_type == lookup.none && self.omit_none {
+                    continue;
+                }
```
Comment (samuelcolvin) [inline src/ser.rs:122]: as above.

File: src/ser.rs [hunk 14]
```diff
@@ -90,30 +98,50 @@ impl<'py> Serialize for SerializePyObject<'py> {
...
                 map.serialize_entry(key, &value)?;
             }
             for (k, v) in array_items {
                 let key = table_key(k)?;
                 let value = self.with_obj(v);
+                let v_ob_type = v.get_type_ptr() as usize;
+                if v_ob_type == lookup.none && self.omit_none {
+                    continue;
+                }
```
Comment (samuelcolvin) [inline src/ser.rs:113]: this isn't required, we've already omitted the `None` cases.

File: src/ser.rs [hunk 15]
```diff
@@ -90,30 +98,50 @@ impl<'py> Serialize for SerializePyObject<'py> {
...
             seq.end()
         } else if ob_type == lookup.tuple {
             let py_tuple: &PyTuple = self.obj.cast_as().map_err(map_py_err)?;
             let mut seq = serializer.serialize_seq(Some(py_tuple.len()))?;
             for element in py_tuple {
+                let v_ob_type = element.get_type_ptr() as usize;
+                if v_ob_type == lookup.none && self.omit_none {
+                    continue;
+                }
                 seq.serialize_element(&self.with_obj(element))?
```
Comment (samuelcolvin) [inline src/ser.rs:145]:
```suggestion
                if self.omit_none && element.is_none() {
                    continue;
                }
                seq.serialize_element(&self.with_obj(element))?
```

Since we're only looking up one type, `.is_none()` will probably be faster than `element.get_type_ptr() as usize == lookup.none`.

File: src/ser.rs [hunk 16]
```diff
@@ -90,30 +98,50 @@ impl<'py> Serialize for SerializePyObject<'py> {
...
                 map.serialize_entry(key, &value)?;
             }
             map.end()
         } else if ob_type == lookup.list {
             let py_list: &PyList = self.obj.cast_as().map_err(map_py_err)?;
             let mut seq = serializer.serialize_seq(Some(py_list.len()))?;
             for element in py_list {
+                let v_ob_type = element.get_type_ptr() as usize;
+                if v_ob_type == lookup.none && self.omit_none {
+                    continue;
```
Comment (samuelcolvin) [inline src/ser.rs:132]: same as tuple case.

**Review by samuelcolvin — COMMENTED** [review]
a few more things to change.

We also need to update the readme.

File: rtoml/__init__.py (hunk 10 above)
Comment (pwwang) [reply rtoml/__init__.py:38]: No, I totally agree using `none_value=None` to omit `None` values.

File: src/ser.rs (hunk 12 above)
Comment (pwwang) [reply src/ser.rs:104]: This could be a nested item, say `{"x": None, "y": [1, 2, None]}`, right?
The exclusion above is excluding `"x"`, but not `None` in "y". Not sure if I understand it correctly.

Comment (samuelcolvin) [reply src/ser.rs:104]: `y` is a list, not `None`.

File: src/ser.rs (hunk 16 above)
Comment (pwwang) [reply src/ser.rs:132]: Sorry, ignore my previous comment.

What confuses me is that we have L126 to handle list (array) items and here we have L106 to handle the list again. I suppose that L106 is handling a list nested deeply in a dict.

Should we consider the None in the nested list here?

Comment (samuelcolvin) [reply src/ser.rs:132]: It's for the order of items, the comment got removed in the the recent updates

```
// https://github.com/alexcrichton/toml-rs/issues/142#issuecomment-278970591
// taken from alexcrichton/toml-rs/blob/ec4e821f3bb081391801e4c00aa90bf66a53562c/src/value.rs#L364-L387
```

**Comment by pwwang** [comment]
As for the readme file, I think this `None`-value handling could be a selling point. Another point is that, in my opinion, you can state that `rtoml` is THE fastest TOML library for python, based on the benchmarking from https://github.com/pwwang/toml-bench

**Comment by samuelcolvin** [comment]
> As for the readme file, I think this `None`-value handling could be a selling point. Another point is that, in my opinion, you can state that `rtoml` is THE fastest TOML library for python, based on the benchmarking from https://github.com/pwwang/toml-bench

I agree on both, we should say "at the time of writing" just in case. Feel free to update here or in another PR.

**Review by samuelcolvin — COMMENTED** [review]
Looking good, I noticed you still need to update `table_key` to use `none_value`, also could you add tests for `{None: 'foobar'}`, both with `none_value='custom-value'` and `none_value=None`?

That and readme updates, and I think we're done. :tada:

**Comment by caniko** [comment]
What happened to this PR?

**Comment by samuelcolvin** [comment]
> What happened to this PR?

I got very busy :smile:.

Fixing now.

**Comment by samuelcolvin** [comment]
Thanks so much for this @pwwang.

**Comment by pwwang** [comment]
Awesome! I also mentioned this feature in https://github.com/pwwang/toml-bench

See sections:
- https://github.com/pwwang/toml-bench/tree/master?tab=readme-ov-file#dumping-none-value
- https://github.com/pwwang/toml-bench/tree/master?tab=readme-ov-file#dumping-key-none-pair
- https://github.com/pwwang/toml-bench/tree/master?tab=readme-ov-file#dumping-list-with-none-value
- https://github.com/pwwang/toml-bench/tree/master?tab=readme-ov-file#loading-none-like-values
